### PR TITLE
Replace encoder with url encoder

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Document IDs are frequently used in HTTP requests, such as `GET /index/_doc/{id}`, where they must be URL-safe to avoid issues with invalid characters. This change ensures that IDs generated by `TimeBasedKOrderedUUIDGenerator` are properly Base64 URL-encoded, free of characters that could break URLs. We also test that no IDs include invalid characters like +, /, or = to guarantee they are fully compliant with URL-safe requirements.

Moreover `TimeBasedKOrderedUUIDGenerator` and `TimeBasedUUIDGenerator` are refactored to allow injection of dependencies which enables us to increase test coverage by including tests for high-throughput scenarios, sequence id overflow and unreliable clocks usage.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
Testing this with the previous encoder actually resulted in failures

```
UUIDTests > testKOrderedUUID FAILED
    java.lang.AssertionError: UUID contains invalid characters: AZMi5TapRhFlU+F2NPIJ
        at __randomizedtesting.SeedInfo.seed([34991382834991A7:16CA98F0E1984993]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.elasticsearch.common.UUIDTests.verifyUUIDIsUrlSafe(UUIDTests.java:245)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.elasticsearch.common.UUIDTests.testKOrderedUUID(UUIDTests.java:51)

REPRODUCE WITH: ./gradlew ":server:test" --tests "org.elasticsearch.common.UUIDTests.testThreadedKOrderedUUID" -Dtests.seed=34991382834991A7 -Dtests.locale=ja-Jpan-JP -Dtests.timezone=America/Belize -Druntime.java=23

UUIDTests > testThreadedKOrderedUUID FAILED
    java.lang.AssertionError: UUID contains invalid characters: AZMi5TapRhFlU+OBp/Xa
        at __randomizedtesting.SeedInfo.seed([34991382834991A7:F891DF8432167CEE]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.elasticsearch.common.UUIDTests.verifyUUIDIsUrlSafe(UUIDTests.java:245)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.elasticsearch.common.UUIDTests.testUUIDThreaded(UUIDTests.java:148)
        at org.elasticsearch.common.UUIDTests.testThreadedKOrderedUUID(UUIDTests.java:63)
 ```

File: server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java [hunk 1]
```diff
@@ -68,6 +68,6 @@ public String getBase64UUID() {
 
         assert buffer.position() == uuidBytes.length;
 
-        return BASE_64_NO_PADDING.encodeToString(uuidBytes);
+        return BASE_64_URL_ENCODER_NO_PADDING.encodeToString(uuidBytes);
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:71]: Let's just use `Strings.BASE_64_NO_PADDING_URL_ENCODER`, it's the same and used elsewhere too.

File: server/src/test/java/org/elasticsearch/common/UUIDTests.java [hunk 2]
```diff
@@ -237,4 +240,13 @@ public void testStringLength() {
     private static int getUnpaddedBase64StringLength(int sizeInBytes) {
         return (int) Math.ceil(sizeInBytes * 4.0 / 3.0);
     }
+
+    private void verifyUUIDIsUrlSafe(final String uuid) {
+        assertFalse("UUID should not contain padding characters: " + uuid, uuid.contains("="));
+        try {
+            BASE_64_URL_DECODER.decode(uuid);
```
Comment (kkrik-es) [inline server/src/test/java/org/elasticsearch/common/UUIDTests.java:247]: I may've missed it, but it'd be nice to test that encoding and decoding returns the original byte array, for random inputs.

**Review by kkrik-es — APPROVED** [review]
Thanks Salvatore. A few nits, nothing major.

File: server/src/test/java/org/elasticsearch/common/UUIDTests.java (hunk 2 above)
Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/common/UUIDTests.java:247]: +1 to such a test

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java [hunk 3]
```diff
@@ -28,18 +28,25 @@
  * The result is a compact base64-encoded string, optimized for efficient compression of the _id field in an inverted index.
  */
 public class TimeBasedKOrderedUUIDGenerator extends TimeBasedUUIDGenerator {
-    private static final Base64.Encoder BASE_64_NO_PADDING = Base64.getEncoder().withoutPadding();
+
+    public TimeBasedKOrderedUUIDGenerator(
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:32]: I did this change here to be able to use dependency injection in unit tests.

File: server/src/test/java/org/elasticsearch/common/TimeBasedUUIDGeneratorTests.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Base64;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+public class TimeBasedUUIDGeneratorTests extends ESTestCase {
```
Comment (salvatore-campagna) [inline server/src/test/java/org/elasticsearch/common/TimeBasedUUIDGeneratorTests.java:22]: I have specific unit tests for both `TimeBasedUUIDGenerator` and `TimeBasedKOrderedUUIDGenerator`...both of them generate UUID based on time and share the same requirements...indeed the only difference between them is the way bytes are ordered.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/119079)|
|✅|8.17|[](https://github.com/elastic/elasticsearch/pull/119080)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**